File: skillmeat/core/similarity.py
"""Similar Artifacts — Detection, Comparison & Consolidation.

Provides SimilarityService for finding, scoring, and ranking similar artifacts
using ArtifactFingerprint + MatchAnalyzer + optional SemanticScorer.

Design notes
------------
- ``ScoreBreakdown`` is a frozen dataclass so it can be used as a dict key or
  in sets without risk of post-construction mutation.
- ``SimilarityResult.match_type`` is derived automatically in ``__post_init__``
  from the composite score — callers never set it directly.
- ``semantic_score`` on ``ScoreBreakdown`` is ``Optional[float]`` because the
  SemanticScorer may be unavailable (not configured) or may time-out; the rest
  of the pipeline continues without it.
- ``SimilarityService`` will be added in a later task once the repository and
  API layers are in place.

Usage
-----
>>> from skillmeat.core.similarity import SimilarityResult, ScoreBreakdown, MatchType
>>> breakdown = ScoreBreakdown(
...     content_score=0.9,
...     structure_score=0.8,
...     metadata_score=0.7,
...     keyword_score=0.6,
... )
>>> result = SimilarityResult(
...     artifact_id="skill:canvas-design",
...     artifact=None,
...     composite_score=0.82,
...     breakdown=breakdown,
... )
>>> result.match_type
<MatchType.NEAR_DUPLICATE: 'near_duplicate'>
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, TYPE_CHECKING

if TYPE_CHECKING:
    pass  # Will add artifact type imports later


class MatchType(str, Enum):
    """Classification of similarity match strength.

    Thresholds
    ----------
    EXACT           Score >= 0.95
    NEAR_DUPLICATE  Score >= 0.80
    SIMILAR         Score >= 0.50
    RELATED         Score >= min_score threshold (anything below 0.50)
    """

    EXACT = "exact"
    NEAR_DUPLICATE = "near_duplicate"
    SIMILAR = "similar"
    RELATED = "related"


@dataclass(frozen=True)
class ScoreBreakdown:
    """Individual score components from similarity analysis.

    All scores are in the range 0.0–1.0.  ``semantic_score`` is ``None``
    when the SemanticScorer is unavailable or timed out; the composite
    score is computed from the remaining components in that case.

    Attributes:
        content_score:   Similarity based on raw file-content hashes.
        structure_score: Similarity based on directory/file-tree structure.
        metadata_score:  Similarity based on title, description, and tags.
        keyword_score:   Similarity based on keyword/TF-IDF match analysis.
        semantic_score:  Embedding-based semantic similarity (optional).
    """

    content_score: float = 0.0
    structure_score: float = 0.0
    metadata_score: float = 0.0
    keyword_score: float = 0.0
    semantic_score: Optional[float] = None  # None when SemanticScorer unavailable/timed out


@dataclass
class SimilarityResult:
    """A single similar artifact with its composite score and breakdown.

    ``match_type`` is derived automatically from ``composite_score`` in
    ``__post_init__`` and must not be provided by callers.

    Attributes:
        artifact_id:     Canonical ``type:name`` identifier for the artifact.
        artifact:        The artifact object (typed as ``object`` until the
                         CachedArtifact import is wired up in a later task).
        composite_score: Weighted aggregate of all score components (0.0–1.0).
        breakdown:       Per-component score breakdown.
        match_type:      Derived strength classification (set in __post_init__).
    """

    artifact_id: str
    artifact: object  # Will be typed as CachedArtifact once available
    composite_score: float
    breakdown: ScoreBreakdown
    match_type: MatchType = field(init=False)

    def __post_init__(self) -> None:
        """Derive match_type from composite_score."""
        if self.composite_score >= 0.95:
            self.match_type = MatchType.EXACT
        elif self.composite_score >= 0.80:
            self.match_type = MatchType.NEAR_DUPLICATE
        elif self.composite_score >= 0.50:
            self.match_type = MatchType.SIMILAR
        else:
            self.match_type = MatchType.RELATED

File: skillmeat/web/types/similarity.ts
/**
 * Similarity Detection Types for SkillMeat
 *
 * These types mirror the backend DTOs for the similar artifacts feature:
 * - SimilarArtifactDTO
 * - SimilarityBreakdownDTO
 * - SimilarArtifactsResponse
 *
 * Field names use snake_case to match JSON response keys exactly.
 * See: GET /api/v1/artifacts/{id}/similar
 *
 * Note: Named SimilarityMatchType (not MatchType) to avoid collision with
 * the MatchType exported by types/discovery.ts (collection membership matching).
 */

/**
 * Match strength classification for a similarity result.
 *
 * - exact: Near-identical content (score ~1.0)
 * - near_duplicate: Very high similarity, minor differences
 * - similar: Meaningful similarity, same domain/purpose
 * - related: Loose similarity, shared concepts or keywords
 */
export type SimilarityMatchType = 'exact' | 'near_duplicate' | 'similar' | 'related';

/**
 * Search scope for similar artifact queries.
 *
 * - collection: Search only the user's personal collection
 * - marketplace: Search marketplace artifacts only
 * - all: Search across both collection and marketplace
 */
export type SimilaritySource = 'collection' | 'marketplace' | 'all';

/**
 * Individual score components from the similarity analysis pipeline.
 *
 * All required scores are in the range [0.0, 1.0].
 * semantic_score is null when a semantic/embedding model is unavailable.
 */
export interface SimilarityBreakdown {
  /** Content similarity score (file content comparison) */
  content_score: number;
  /** Structure similarity score (artifact structure/schema comparison) */
  structure_score: number;
  /** Metadata similarity score (name, description, tags comparison) */
  metadata_score: number;
  /** Keyword/TF-IDF similarity score */
  keyword_score: number;
  /** Semantic similarity score via embeddings; null if unavailable */
  semantic_score: number | null;
}

/**
 * A similar artifact with its similarity score and detailed breakdown.
 *
 * Mirrors SimilarArtifactDTO from the backend.
 */
export interface SimilarArtifact {
  /** Unique identifier of the similar artifact */
  artifact_id: string;
  /** Display name of the similar artifact */
  name: string;
  /** Type of the artifact (skill, command, agent, mcp, hook) */
  artifact_type: string;
  /** Source repository or path; null for locally-created artifacts */
  source: string | null;
  /** Overall similarity score in the range [0.0, 1.0] */
  composite_score: number;
  /** Match strength classification */
  match_type: SimilarityMatchType;
  /** Individual score components from the similarity analysis */
  breakdown: SimilarityBreakdown;
}

/**
 * Response envelope for the GET /api/v1/artifacts/{id}/similar endpoint.
 *
 * Mirrors SimilarArtifactsResponse from the backend.
 * Items are ordered by composite_score descending.
 */
export interface SimilarArtifactsResponse {
  /** Canonical identifier of the source artifact that was queried */
  artifact_id: string;
  /** Similar artifacts ordered by composite score descending */
  items: SimilarArtifact[];
  /** Total number of similar artifacts returned */
  total: number;
}

/**
 * Options for the useSimilarArtifacts hook.
 *
 * All fields are optional; the hook applies sensible defaults.
 */
export interface SimilarArtifactsOptions {
  /** Maximum number of similar artifacts to return */
  limit?: number;
  /** Minimum composite score threshold (0.0–1.0) */
  minScore?: number;
  /** Search scope: collection, marketplace, or all */
  source?: SimilaritySource;
  /** Whether the query is enabled (default: true) */
  enabled?: boolean;
}

// ============================================================================
// Similarity Settings Types
// ============================================================================

/**
 * Score thresholds that classify a similarity result into a band.
 *
 * The score bands are applied in descending order:
 *   score >= high    → "high" match (strong duplicate / near-identical)
 *   score >= partial → "partial" match (meaningful overlap)
 *   score >= low     → "low" match (some shared concepts)
 *   score >= floor   → surfaced at all (below this, results are suppressed)
 *
 * All values are in the range [0.0, 1.0].
 *
 * Mirrors SimilarityThresholdsDTO from the backend.
 * See: GET /api/v1/settings/similarity
 */
export interface SimilarityThresholds {
  /** Minimum score for a "high" match band (default: 0.80) */
  high: number;
  /** Minimum score for a "partial" match band (default: 0.55) */
  partial: number;
  /** Minimum score for a "low" match band (default: 0.35) */
  low: number;
  /** Absolute minimum score — results below this are excluded (default: 0.20) */
  floor: number;
}

/**
 * CSS color strings assigned to each similarity score band.
 *
 * Used to render score badges and highlight rows in the Similar Artifacts tab.
 * Values must be valid CSS color strings (hex, rgb, hsl, named colors, etc.).
 *
 * Mirrors SimilarityColorsDTO from the backend.
 * See: GET /api/v1/settings/similarity
 */
export interface SimilarityColors {
  /** Color for the "high" score band (default: "#22c55e" — green-500) */
  high: string;
  /** Color for the "partial" score band (default: "#eab308" — yellow-500) */
  partial: string;
  /** Color for the "low" score band (default: "#f97316" — orange-500) */
  low: string;
}

/**
 * Combined similarity settings response from GET /api/v1/settings/similarity.
 *
 * Bundles thresholds and colors into a single payload so consumers
 * can read both in a single query.
 */
export interface SimilaritySettings {
  thresholds: SimilarityThresholds;
  colors: SimilarityColors;
}
